app/src/main/java/com/sschuraytz/charcoaldrawing/DrawingView.java

```diff
@@ -135,4 +110,18 @@ public void setEraseMode() {
     public void setDrawingMode() {
         currentTool = charcoalTool;
     }
-}
+
+    protected void undo() {
```

Comment:
This doesn't work, does it?
It seems to me that you keep pushing and popping the same Bitmap onto the Stack

Reply:
It doesn't work _the way it should_.
Eventually, if I keep clicking undo, it undoes everything, and redo redoes everything.
But I don't think I'm **just** pushing/popping the same Bitmap. There are 2 stacks - "currentStack" and "undoneStack". They are essentially opposites - so the pop from one is pushed to the other.

Reply:
But the *bitmap* you are pushing is the same object. For this to work you need to have different bitmaps every time you draw.

Reply:
A stack of bitmaps doesn't save each pushed bitmap as a different bitmap/ a different **state** of the bitmap (even though each push originates from the same bitmap)? 

Is this a bad implementation if I need to create a different bitmap every time the user draws?

Reply:
It's the same Bitmap each time. So when you draw onto one, you draw on all of them.

It's not a bad architecture to have multiple bitmaps. Try it. We can discuss it more tomorrow.

Also I'd focus on getting undo to week before doing redo.